lib/scholar/neighbors/knn_classifier.ex

```diff
@@ -0,0 +1,245 @@
+
+        * `:distance` - weight points by the inverse of their distance. in this case, closer neighbors of
+          a query point will have a greater influence than neighbors which are further away.
+      """
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+
+  @doc """
+  Fits a k-NN classifier model.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Examples
+
+      iex> x = Nx.tensor([[1, 2], [2, 3], [3, 4], [4, 5], [5, 6]])
+      iex> y = Nx.tensor([0, 0, 0, 1, 1])
+      iex> model = Scholar.Neighbors.KNNClassifier.fit(x, y, num_neighbors: 3, num_classes: 2)
+      iex> model.algorithm
+      Scholar.Neighbors.BruteKNN.fit(x, num_neighbors: 3)
+      iex> model.labels
+      Nx.tensor([0, 0, 0, 1, 1])
+  """
+  deftransform fit(x, y, opts) do
+    if Nx.rank(x) != 2 do
+      raise ArgumentError,
+            "expected x to have shape {num_samples, num_features},
+             got tensor with shape: #{inspect(Nx.shape(x))}"
+    end
+
+    if Nx.rank(y) != 1 and Nx.axis_size(x, 0) == Nx.axis_size(y, 0) do
+      raise ArgumentError,
+            "expected y to have shape {num_samples},
+             got tensor with shape: #{inspect(Nx.shape(y))}"
+    end
+
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+
+    {algorithm_name, algorithm_opts} =
+      if is_atom(opts[:algorithm]) do
+        {opts[:algorithm], []}
+      else
+        opts[:algorithm]
+      end
+
+    knn_module =
+      case algorithm_name do
+        :brute ->
+          Scholar.Neighbors.BruteKNN
+
+        :kd_tree ->
+          Scholar.Neighbors.KDTree
+
+        :random_projection_forest ->
+          Scholar.Neighbors.RandomProjectionForest
+
+        knn_module when is_atom(knn_module) ->
+          knn_module
+
+        _ ->
+          raise ArgumentError,
+                """
+                not supported
+                """
+      end
+
+    # TODO: Maybe raise an error if :num_neighbors or :metric is already in algorithm_opts?
+
+    algorithm_opts = Keyword.put(algorithm_opts, :num_neighbors, opts[:num_neighbors])
+    algorithm_opts = Keyword.put(algorithm_opts, :metric, opts[:metric])
+
+    algorithm = knn_module.fit(x, algorithm_opts)
+
+    %__MODULE__{
+      algorithm: algorithm,
+      num_classes: opts[:num_classes],
+      labels: y,
+      weights: opts[:weights]
+    }
+  end
+
+  @doc """
+  Makes predictions using a k-NN classifier model.
+
+  ## Examples
+
+      iex> x_train = Nx.tensor([[1, 2], [2, 3], [3, 4], [4, 5], [5, 6]])
+      iex> y_train = Nx.tensor([0, 0, 0, 1, 1])
+      iex> model = Scholar.Neighbors.KNNClassifier.fit(x_train, y_train, num_neighbors: 3, num_classes: 2)
+      iex> x_test = Nx.tensor([[1, 3], [4, 2], [3, 6]])
+      iex> Scholar.Neighbors.KNNClassifier.predict(model, x_test)
+      Nx.tensor([0, 0, 1])
+  """
+  deftransform predict(model, x) do
+    knn_module = model.algorithm.__struct__
```

Comment:
This doesn't work inside `defn`. Not sure how to resolve it.

Reply:
What is the issue with this code? It looks fine as is to me.

Reply:
I wanted to put `knn_module.predict(model.algorithm, x)` inside `defn` (e.g. `predict_n`), but the line above won't work (invalid numerical expression error as far as I remember).

Reply:
I see. In any case, this code looks fine to me, I can play with it later on and see if I can improve the defn error.

Reply:
Apparently, it's not this line that is causing the problem. The following line is giving me `invalid numerical expression` error
```
{neighbors, distances} = algorithm_module.predict(model.algorithm, x)
```
I am not sure why, but it probably has to do something with how `defn` works. Could you maybe have a look?

Reply:
Ok, it makes sense it doesn't work, because `defn` then cannot guarantee `algorithm_module.predict` is a `defn` at compile-time. We can likely relax it though but that shouldn't block it here. My suggestion is to make `predict` a `defn predict` and move only the invocation to transform:

```
defn predict(model, x) do
  {neighbors, distances} = predict_apply(model.algorithm, x)
  ...
end

deftransformp predict_apply(algo, x) do
  algo.__struct__.predict(algo, x)
end
```

We can probably do something similar for calling `fit` on init.

Reply:
FWIW, it is not strictly necessary for calling `fit`, because in `fit` the call to `algo.fit` is literally the only `defn` that we invoke, but it probably makes sense to do it as a best practice anyway.

Reply:
> FWIW, it is not strictly necessary for calling `fit`, because in `fit` the call to `algo.fit` is literally the only `defn` that we invoke, but it probably makes sense to do it as a best practice anyway.

I didn't use `defn` for `fit` exactly for the reason you wrote above. Are you suggesting to do it anyway? I.e. to move `algorithm_module.fit(x, algorithm_opts)` inside `defn`?

Reply:
It is up to you. I am fine either way. :)

Reply:
I think it just makes the code slightly harder to read, so I would avoid doing it for `fit`. :)